torch/distributed/_functional_collectives.py

```diff
@@ -115,13 +115,17 @@ def cleanup(self):
                 del data_ptr_to_work[ptr]
 
 
-def _register_tensor_work(tensor_or_list, work):
-    reg = _WaitRegistration(work)
+def _register_tensor_work(tensor_or_list, work_or_list):
     if not isinstance(tensor_or_list, list):
         tensor_or_list = [tensor_or_list]
-    for tensor in tensor_or_list:
-        reg._register(tensor)
-
+    if not isinstance(work_or_list, list):
+        reg = _WaitRegistration(work_or_list)
```

Comment:
for tensor_list case, we should just recursively call into this function?

Reply:
We want a single _WaitRegistration instance for all tensors.